### PR TITLE
Add fastp component

### PR DESCRIPTION
**Pull request by rcannood**
## Description

* Add fastp component

## Issue ticket number

Closes #xxxx 

## Checklist before requesting a review

- [x] I have performed a self-review of my code

- [x] Conforms to the [Contributing guidelines](https://github.com/viash-hub/base/blob/main/CONTRIBUTING.md)

- [x] Proposed changes are described in the [CHANGELOG.md](https://github.com/viash-hub/base/blob/main/CHANGELOG.md)

- [x] I have tested my code with `viash ns test --parallel -q <name or namespace>`

- Check the correct box. Does this PR contain:
  - [ ] Breaking changes
  - [x] New functionality
  - [ ] Major changes
  - [ ] Minor changes
  - [ ] Documentation
  - [ ] Bug fixes

### CODE REVIEW

File: src/fastp/config.vsh.yaml [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+          example: failed.fq.gz
+          direction: output
+        - name: --overlapped_out
+          type: file
+          description: |
+            For each read pair, output the overlapped region if it has no any mismatched base.
+          direction: output
+    - name: Report output arguments
+      arguments:
+        - name: --json
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:100]: Add defaults?

File: src/fastp/config.vsh.yaml [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+          alternatives: [-j]
+          type: file
+          description: |
+            The json format report file name
+          example: out.json
+          direction: output
+        - name: --html
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:107]: Add defaults?

File: src/fastp/config.vsh.yaml [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+          alternatives: ["-M"]
+          type: integer
+          description: |
+            The mean quality requirement option shared by cut_front, cut_tail or cut_sliding. Range: 1~36 default: 20 (Q20)
+          example: 20
+        - name: --cut_front_window_size
+          type: integer
+          description: |
+            The window size option of cut_front, default to cut_window_size if not specified.
+          example: 4
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:341]:
```suggestion
          example: 4
          min: 1
```

File: src/fastp/config.vsh.yaml [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+            Indicate the input is using phred64 scoring (it'll be converted to phred33, so the output will still be phred33)
+    - name: Additional output arguments
+      description: Affects how the output is written.
+      arguments:
+        - name: --compression
+          alternatives: ["-z"]
+          type: integer
+          description: |
+            Compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 4.
+          example: 4
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:246]:
```suggestion
          example: 4
          min: 1
          max: 9
```

File: src/fastp/config.vsh.yaml [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+          example: 0
+        - name: --trim_tail2
+          alternatives: [-T]
+          type: integer
+          description: |
+            Trimming how many bases in tail for read2, default is 0.
+          example: 0
+        - name: --max_len2
+          alternatives: [-B]
+          type: integer
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:201]:
```suggestion
          type: integer
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+          example: 0
+        - name: --trim_tail1
+          alternatives: [-t]
+          type: integer
+          description: |
+            Trimming how many bases in tail for read1, default is 0.
+          example: 0
+        - name: --max_len1
+          alternatives: [-b]
+          type: integer
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:184]:
```suggestion
          type: integer
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+          alternatives: [-V]
+          type: boolean_true
+          description: Output verbose log information (i.e. when every 1M reads are processed).
+    - name: Processing arguments
+      arguments:
+        - name: --reads_to_process
+          type: long
+          description: |
+            Specify how many reads/pairs to be processed. Default 0 means process all reads.
+          example: 1000000
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:263]:
```suggestion
          example: 1000000
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+      arguments:
+        - name: --dedup
+          type: boolean_true
+          description: |
+            Enable deduplication to drop the duplicated reads/pairs
+        - name: --dup_calc_accuracy
+          type: integer
+          description: |
+            Accuracy level to calculate duplication (1~6). Higher level uses more memory (1G, 2G, 4G, 8G, 16G, 24G). Default 1 for no-dedup mode, and 3 for dedup mode.
+          example: 3
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:274]:
```suggestion
          example: 3
          choices: [0, 1, 2, 3, 4, 5, 6]
```
or 
```suggestion
          example: 3
          min: 0
          max: 6
```

File: src/fastp/config.vsh.yaml [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --trim_poly_g
+          alternatives: [-g]
+          type: boolean_true
+          description: |
+            Force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data
+        - name: --poly_g_min_len
+          type: integer
+          description: |
+            The minimum length to detect polyG in the read tail. 10 by default.
+          example: 10
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:290]:
```suggestion
          example: 10
          min: 1
```

File: src/fastp/config.vsh.yaml [hunk 10]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --trim_poly_x
+          alternatives: [-x]
+          type: boolean_true
+          description: |
+            Enable polyX trimming in 3' ends.
+        - name: --poly_x_min_len
+          type: integer
+          description: |
+            The minimum length to detect polyX in the read tail. 10 by default.
+          example: 10
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:307]:
```suggestion
          example: 10
          min: 1
```

File: src/fastp/config.vsh.yaml [hunk 11]
```diff
@@ -0,0 +1,541 @@
...
+          alternatives: ["-r"]
+          type: integer
+          description: |
+            Move a sliding window from front to tail, if meet one window with mean quality < threshold, drop the bases in the window and the right part, and then stop.
+        - name: --cut_window_size
+          alternatives: ["-W"]
+          type: integer
+          description: |
+            The window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4.
+          example: 4
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:330]:
```suggestion
          example: 4
          min: 1
```

File: src/fastp/config.vsh.yaml [hunk 12]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --cut_mean_quality
+          alternatives: ["-M"]
+          type: integer
+          description: |
+            The mean quality requirement option shared by cut_front, cut_tail or cut_sliding. Range: 1~36 default: 20 (Q20)
+          example: 20
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:336]:
```suggestion
          example: 20
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 13]
```diff
@@ -0,0 +1,541 @@
...
+          alternatives: [-Q]
+          type: boolean_true
+          description: |
+            Quality filtering is enabled by default. If this option is specified, quality filtering is disabled.
+        - name: --qualified_quality_phred
+          alternatives: [-q]
+          type: integer
+          description: |
+            The quality value that a base is qualified. Default 15 means phred quality >=Q15 is qualified.
+          example: 15
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:379]:
```suggestion
          example: 15
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 14]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --cut_right_window_size
+          type: integer
+          description: |
+            The window size option of cut_right, default to cut_window_size if not specified.
+          example: 4
+        - name: --cut_right_mean_quality
+          type: integer
+          description: |
+            The mean quality requirement option of cut_right, default to cut_mean_quality if not specified.
+          example: 20
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:366]:
```suggestion
          example: 20
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 15]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --cut_tail_mean_quality
+          type: integer
+          description: |
+            The mean quality requirement option of cut_tail, default to cut_mean_quality if not specified.
+          example: 20
+        - name: --cut_right_window_size
+          type: integer
+          description: |
+            The window size option of cut_right, default to cut_window_size if not specified.
+          example: 4
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:361]:
```suggestion
          example: 4
          min: 1
```

File: src/fastp/config.vsh.yaml [hunk 16]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --cut_tail_window_size
+          type: integer
+          description: |
+            The window size option of cut_tail, default to cut_window_size if not specified.
+          example: 4
+        - name: --cut_tail_mean_quality
+          type: integer
+          description: |
+            The mean quality requirement option of cut_tail, default to cut_mean_quality if not specified.
+          example: 20
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:356]:
```suggestion
          example: 20
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 17]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --cut_front_mean_quality
+          type: integer
+          description: |
+            The mean quality requirement option of cut_front, default to cut_mean_quality if not specified.
+          example: 20
+        - name: --cut_tail_window_size
+          type: integer
+          description: |
+            The window size option of cut_tail, default to cut_window_size if not specified.
+          example: 4
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:351]:
```suggestion
          example: 4
          min: 1
```

File: src/fastp/config.vsh.yaml [hunk 18]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --cut_front_window_size
+          type: integer
+          description: |
+            The window size option of cut_front, default to cut_window_size if not specified.
+          example: 4
+        - name: --cut_front_mean_quality
+          type: integer
+          description: |
+            The mean quality requirement option of cut_front, default to cut_mean_quality if not specified.
+          example: 20
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:346]:
```suggestion
          example: 20
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 19]
```diff
@@ -0,0 +1,541 @@
...
+          type: integer
+          description: |
+            How many percents of bases are allowed to be unqualified (0~100). Default 40 means 40%.
+          example: 40
+        - name: --n_base_limit
+          alternatives: [-n]
+          type: integer
+          description: |
+            If one read's number of N base is >n_base_limit, then this read/pair is discarded. Default is 5.
+          example: 5
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:391]:
```suggestion
          example: 5
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 20]
```diff
@@ -0,0 +1,541 @@
...
+          type: integer
+          description: |
+            The quality value that a base is qualified. Default 15 means phred quality >=Q15 is qualified.
+          example: 15
+        - name: --unqualified_percent_limit
+          alternatives: [-u]
+          type: integer
+          description: |
+            How many percents of bases are allowed to be unqualified (0~100). Default 40 means 40%.
+          example: 40
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:385]:
```suggestion
          example: 40
          min: 0
          max: 100
```

File: src/fastp/config.vsh.yaml [hunk 21]
```diff
@@ -0,0 +1,541 @@
...
+          type: integer
+          description: |
+            If one read's number of N base is >n_base_limit, then this read/pair is discarded. Default is 5.
+          example: 5
+        - name: --average_qual
+          alternatives: [-e]
+          type: integer
+          description: |
+            If one read's average quality score <avg_qual, then this read/pair is discarded. Default 0 means no requirement.
+          example: 0
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:397]:
```suggestion
          example: 0
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 22]
```diff
@@ -0,0 +1,541 @@
...
+          alternatives: [-L]
+          type: boolean_true
+          description: |
+            Length filtering is enabled by default. If this option is specified, length filtering is disabled.
+        - name: --length_required
+          alternatives: [-l]
+          type: integer
+          description: |
+            Reads shorter than length_required will be discarded, default is 15.
+          example: 15
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:410]:
```suggestion
          example: 15
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 23]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --length_limit
+          type: integer
+          description: |
+            Reads longer than length_limit will be discarded, default 0 means no limitation.
+          example: 0
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:415]:
```suggestion
          example: 0
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 24]
```diff
@@ -0,0 +1,541 @@
...
+          alternatives: [-y]
+          type: boolean_true
+          description: |
+            Enable low complexity filter. The complexity is defined as the percentage of base that is different from its next base (base[i] != base[i+1]).
+        - name: --complexity_threshold
+          alternatives: [-Y]
+          type: integer
+          description: |
+            The threshold for low complexity filter (0~100). Default is 30, which means 30% complexity is required.
+          example: 30
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:428]:
```suggestion
          example: 30
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 25]
```diff
@@ -0,0 +1,541 @@
...
+            Specify a file contains a list of barcodes of index1 to be filtered out, one barcode per line.
+        - name: --filter_by_index2
+          type: file
+          description: |
+            Specify a file contains a list of barcodes of index2 to be filtered out, one barcode per line.
+        - name: --filter_by_index_threshold
+          type: integer
+          description: |
+            The allowed difference of index barcode for index filtering, default 0 means completely identical.
+          example: 0
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:443]:
```suggestion
          example: 0
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 26]
```diff
@@ -0,0 +1,541 @@
...
+        - type: boolean_true
+          name: --correction
+          alternatives: [-c]
+          description: |
+            Enable base correction in overlapped regions (only for PE data), default is disabled.
+        - name: --overlap_len_require
+          type: integer
+          description: |
+            The minimum length to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. 30 by default.
+          example: 30
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:455]:
```suggestion
          example: 30
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 27]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --overlap_diff_limit
+          type: integer
+          description: |
+            The maximum number of mismatched bases to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. 5 by default.
+          example: 5
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:460]:
```suggestion
          example: 5
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 28]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --overlap_diff_percent_limit
+          type: integer
+          description: |
+            The maximum percentage of mismatched bases to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. Default 20 means 20%.
+          example: 20
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:465]:
```suggestion
          example: 20
          min: 0
          max: 100
```

File: src/fastp/config.vsh.yaml [hunk 29]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --umi_loc
+          type: string
+          description: |
+            Specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none.
+          choices: [index1, index2, read1, read2, per_index, per_read]
+        - name: --umi_len
+          type: integer
+          description: |
+            Iif the UMI is in read1/read2, its length should be provided.
+          example: 0
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:482]:
```suggestion
          example: 0
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 30]
```diff
@@ -0,0 +1,541 @@
...
+            Enable unique molecular identifier (UMI) preprocessing.
+        - name: --umi_loc
+          type: string
+          description: |
+            Specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none.
+          choices: [index1, index2, read1, read2, per_index, per_read]
+        - name: --umi_len
+          type: integer
+          description: |
+            Iif the UMI is in read1/read2, its length should be provided.
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:481]:
```suggestion
            If the UMI is in read1/read2, its length should be provided.
```

File: src/fastp/config.vsh.yaml [hunk 31]
```diff
@@ -0,0 +1,541 @@
...
+          example: 0
+        - name: --umi_prefix
+          type: string
+          description: |
+            If specified, an underline will be used to connect prefix and UMI (i.e. prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default.
+        - name: --umi_skip
+          type: integer
+          description: |
+            If the UMI is in read1/read2, fastp can skip several bases following UMI, default is 0.
+          example: 0
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:491]:
```suggestion
          example: 0
          min: 0
```

File: src/fastp/config.vsh.yaml [hunk 32]
```diff
@@ -0,0 +1,541 @@
...
+        - name: --overrepresentation_analysis
+          alternatives: [-p]
+          type: boolean_true
+          description: |
+            Enable overrepresentation analysis.
+        - name: --overrepresentation_sampling
+          type: integer
+          description: |
+            One in (--overrepresentation_sampling) reads will be computed for overrepresentation analysis (1~10000), smaller is slower, default is 20.
+          example: 20
```
Comment (DriesSchaumont) [inline src/fastp/config.vsh.yaml:507]:
```suggestion
          example: 20
          min: 1
```

File: src/fastp/test.sh [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+  --in1 "$meta_resources_dir/test_data/pe/a.1.fastq" \
+  --in2 "$meta_resources_dir/test_data/pe/a.2.fastq" \
+  --out1 "trimmed_1.fastq" \
+  --out2 "trimmed_2.fastq" \
+  --failed_out "failed.fastq" \
+  --json "report.json" \
+  --html "report.html" \
+  --adapter_sequence ACGGCTAGCTA \
+  --adapter_sequence_r2 AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC \
+  --merge \
```
Comment (DriesSchaumont) [inline src/fastp/test.sh:36]: Test without `--merge`?

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]

File: src/fastp/config.vsh.yaml (hunk 1 above)
Comment (rcannood) [reply src/fastp/config.vsh.yaml:100]: A user may or may not want a report as an output, so I'd say not adding defaults for --json and --html is the way to go

File: src/fastp/config.vsh.yaml [hunk 34]
```diff
@@ -0,0 +1,541 @@
...
+    - report JSON format result for further interpreting.
+    - visualize quality control and filtering results on a single HTML page (like FASTQC but faster and more informative).
+    - split the output to multiple files (0001.R1.gz, 0002.R1.gz...) to support parallel processing. Two modes can be used, limiting the total split file number, or limitting the lines of each split file.
+    - support long reads (data from PacBio / Nanopore devices).
+    - support reading from STDIN and writing to STDOUT
+    - support interleaved input
+    - support ultra-fast FASTQ-level deduplication
+  info:
+    keywords: [RNA-Seq, Trimming, Quality control]
+    repository: https://github.com/OpenGene/fast
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/fastp/config.vsh.yaml:25]: Thnaks!

File: src/fastp/config.vsh.yaml [hunk 35]
```diff
@@ -0,0 +1,541 @@
+functionality:
+  name: fastp
+  description: |
+    An ultra-fast all-in-one FASTQ preprocessor (QC/adapters/trimming/filtering/splitting/merging...).
+
+    Features:
+
+    - comprehensive quality profiling for both before and after filtering data (quality curves, base contents, KMER, Q20/Q30, GC Ratio, duplication, adapter contents...)
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/fastp/config.vsh.yaml:8]: Done!

File: src/fastp/config.vsh.yaml (hunk 2 above)
Comment (rcannood) [reply src/fastp/config.vsh.yaml:107]: see above

File: src/fastp/config.vsh.yaml (hunk 1 above)
Comment (DriesSchaumont) [reply src/fastp/config.vsh.yaml:100]: Ok, just wanted to make sure that we are deviating from what fastp does by default.
```
 -j, --json                           the json format report file name (string [=fastp.json])
 -h, --html                           the html format report file name (string [=fastp.html])
```

File: src/fastp/test.sh (hunk 33 above)
Comment (rcannood) [reply src/fastp/test.sh:36]: done!

**Review by DriesSchaumont — APPROVED** [review]

**Pull request merged**